chain/chain/src/flat_storage_resharder.rs

```diff
@@ -460,6 +476,180 @@ impl FlatStorageResharder {
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        if height > chain_final_head.height {
+            info!(target = "resharding", ?height, chain_final_height = ?chain_final_head.height, "flat head beyond chain final tip: postponing flat storage shard catchup task");
+            self.scheduler.send(ReshardingRequest::FlatStorageShardCatchup {
+                resharder: self.clone(),
+                shard_uid,
+                flat_head_block_hash,
+            });
+            return FlatStorageReshardingTaskStatus::Cancelled;
+        }
+
+        // If the flat head is not in the canonical chain this task has failed.
+        match chain_store.get_block_hash_by_height(height) {
+            Ok(hash) => {
+                if hash != flat_head_block_hash {
+                    error!(target: "resharding", ?shard_uid, ?height, ?flat_head_block_hash, ?hash, "flat head not in canonical chain!");
+                    return FlatStorageReshardingTaskStatus::Failed;
+                }
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, ?height, "can't find block hash by height");
+                return FlatStorageReshardingTaskStatus::Failed;
+            }
+        }
+
+        // Apply deltas and then create the flat storage.
+        let apply_result =
+            self.shard_catchup_apply_deltas(shard_uid, flat_head_block_hash, chain_store);
+        let Ok((num_batches_done, flat_head)) = apply_result else {
+            error!(target: "resharding", ?shard_uid, err = ?apply_result.unwrap_err(), "flat storage shard catchup delta application failed!");
+            return FlatStorageReshardingTaskStatus::Failed;
+        };
+        match self.shard_catchup_finalize_storage(shard_uid, &flat_head) {
+            Ok(_) => {
+                let task_status = FlatStorageReshardingTaskStatus::Successful { num_batches_done };
+                info!(target: "resharding", ?shard_uid, ?task_status, "flat storage shard catchup task finished");
+                // At this point we can trigger the reload of memtries.
+                self.scheduler.send(ReshardingRequest::MemtrieReload { shard_uid });
+                task_status
+            }
+            Err(err) => {
+                error!(target: "resharding", ?shard_uid, ?err, "flat storage shard catchup finalize failed!");
+                FlatStorageReshardingTaskStatus::Failed
+            }
+        }
+    }
+
+    /// Applies flat storage deltas in batches on a shard that is in catchup status.
+    ///
+    /// Returns the number of delta batches applied and the final tip of the flat storage.
+    fn shard_catchup_apply_deltas(
+        &self,
+        shard_uid: ShardUId,
+        mut flat_head_block_hash: CryptoHash,
+        chain_store: &ChainStore,
+    ) -> Result<(usize, Tip), Error> {
+        const CATCH_UP_BLOCKS: u32 = 50;
```

Comment:
Can you make it be part of the config? 

Cool that we have rate limiting for this. Why number of blocks? It makes perfect sense but I'm just curious what's your thinking.

Reply:
It's the same logic used in `flat_storage_creator` to limit memory usage.

Sure, I can add it in the config, likely in another PR.